core/src/main/java/org/sonarsource/sonarlint/core/websocket/WebSocketService.java

```diff
@@ -45,135 +48,217 @@
+  private final boolean shouldEnableWebSockets;
+  private String connectionIdUsedToCreateConnection;
   private final ConnectionConfigurationRepository connectionConfigurationRepository;
   private final ConfigurationRepository configurationRepository;
   private final ConnectionAwareHttpClientProvider connectionAwareHttpClientProvider;
   protected SonarCloudWebSocket sonarCloudWebSocket;
   private final EventDispatcher eventRouter;
 
   public WebSocketService(SonarLintClient client, ConnectionConfigurationRepository connectionConfigurationRepository, ConfigurationRepository configurationRepository,
-    ConnectionAwareHttpClientProvider connectionAwareHttpClientProvider, TelemetryServiceImpl telemetryService) {
+    ConnectionAwareHttpClientProvider connectionAwareHttpClientProvider, TelemetryServiceImpl telemetryService, InitializeParams params) {
     this.connectionConfigurationRepository = connectionConfigurationRepository;
     this.configurationRepository = configurationRepository;
     this.connectionAwareHttpClientProvider = connectionAwareHttpClientProvider;
+    this.shouldEnableWebSockets = params.getFeatureFlags().shouldManageSmartNotifications();
     this.eventRouter = new EventDispatcher()
       .dispatch(QualityGateChangedEvent.class, new ShowSmartNotificationOnQualityGateChangedEvent(client, configurationRepository, telemetryService));
   }
 
-  protected void refreshConnectionIfNeeded() {
+  protected void reopenConnectionOnClose() {
     var connectionId = connectionIdsInterestedInNotifications.stream().findFirst().orElse(null);
     if (this.sonarCloudWebSocket != null && connectionId != null) {
       // If connection already exists, close it and create new one before it expires on its own
-      closeSocket();
-      createConnectionIfNeeded(connectionId);
-      resubscribeAll();
+      reopenConnection(connectionId);
     }
   }
 
   @Subscribe
   public void handleEvent(BindingConfigChangedEvent bindingConfigChangedEvent) {
-    var newProjectKey = bindingConfigChangedEvent.getNewConfig().getSonarProjectKey();
-    var configScopeId = bindingConfigChangedEvent.getNewConfig().getConfigScopeId();
-    String connectionId = bindingConfigChangedEvent.getNewConfig().getConnectionId();
-    var connection = connectionId != null ? connectionConfigurationRepository.getConnectionById(connectionId) : null;
-    var isSonarCloudConnection = connection != null && connection.getKind().equals(ConnectionKind.SONARCLOUD);
-    var projectKey = subscribedProjectKeysByConfigScopes.remove(bindingConfigChangedEvent.getNewConfig().getConfigScopeId());
-    if (projectKey != null && !subscribedProjectKeysByConfigScopes.containsValue(projectKey)) {
-      unsubscribe(projectKey);
-    }
-    if (newProjectKey != null && isSonarCloudConnection && !connection.isDisableNotifications()) {
-      createConnectionIfNeeded(connectionId);
-      subscribe(configScopeId, newProjectKey);
+    if (!shouldEnableWebSockets) {
```

Comment:
this check looks a bit annoying in every method. Is it only for limiting which tests initialize websocket connection?

Reply:
I don't like it either. The service works exclusively in reaction to events, so the only way I found to respect the flag is to check at the beginning of each `handle` method. The event bus implementation only allows us to use annotations to subscribe. Maybe we should find another implementation that lets us subscribe programmatically, this way we could have the check in a single place